File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.py
from collections import deque

def bfs(n, m, grid):
    # 상, 하, 좌, 우 방향 정의
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    # 방문 배열 (벽을 부쉈을 때와 아닐 때 두 개의 차원으로 관리)
    visited = [[[False] * 2 for _ in range(m)] for _ in range(n)]
    
    # BFS 큐 (x, y, 현재 거리, 벽 부숨 여부)
    queue = deque([(0, 0, 1, 0)])  # (row, col, distance, wall_broken)
    visited[0][0][0] = True  # 시작 위치 방문 처리 (벽 부수지 않음)

    while queue:
        x, y, dist, broken = queue.popleft()
        
        # 도착 지점에 도달하면 거리 반환
        if x == n - 1 and y == m - 1:
            return dist

        for dx, dy in directions:
            nx, ny = x + dx, y + dy

            # 맵을 벗어나지 않는지 확인
            if 0 <= nx < n and 0 <= ny < m:
                # 빈 칸이고, 아직 방문하지 않았다면
                if grid[nx][ny] == 0 and not visited[nx][ny][broken]:
                    visited[nx][ny][broken] = True
                    queue.append((nx, ny, dist + 1, broken))

                # 벽이고, 아직 벽을 부순 적이 없다면
                elif grid[nx][ny] == 1 and broken == 0 and not visited[nx][ny][1]:
                    visited[nx][ny][1] = True
                    queue.append((nx, ny, dist + 1, 1))  # 벽을 부순 상태로 추가

    return -1  # 도달할 수 없는 경우

# 입력 받기
n, m = map(int, input().split())
grid = [list(map(int, input().strip())) for _ in range(n)]

# 결과 출력
print(bfs(n, m, grid))

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.js
const readline = require("readline")
const rl = readline.createInterface({
    input:process.stdin,
    output:process.stdout
})


const lines = []


rl.on("line", (line)=>lines.push(line)).on("close", ()=>{
    const [N,M] = lines[0].split(" ").map(Number)
    const arr = lines.slice(1).map(item => item.split("").map(Number))
    const visited = []
    for(let i=0;i<N;i++){
        const row = []
        for(let j=0;j<M;j++){
            row.push([0,0])
        }
        visited.push(row)
    }
    const que = [[0,0,0]] // (r,c, broken)
    visited[0][0][0]=1
    let answer = -1
    let head = 0
    // 0은 이동할 수 있는 곳을 나타내고, 1은 이동할 수 없는 벽이 있는 곳
    // console.log(N,M,arr)
    // console.log(visited)
    const directions = [
        [1,0],[-1,0],[0,1],[0,-1]
    ]
    while(que.length>head){
        // console.log(head, que)
        const [r,c,broken] = que[head++]
        if(r===N-1 && c==M-1){
            answer = (visited[r][c][broken])
            break
        }
        for(let [dr, dc] of directions){
            const [nr, nc] = [r+dr, c+dc]
            if(0>nr || nr>=N || 0>nc || nc>=M){continue}
            // 벽이 아님
            if(arr[nr][nc]===0 &&visited[nr][nc][broken] ==0 ){
                visited[nr][nc][broken] = visited[r][c][broken]+1
                que.push([nr, nc, broken])
            } // 벽임 && 아직 벽 뿌신 적 없음
            else if(arr[nr][nc]===1 && broken<1 &&visited[nr][nc][broken] ==0 ){
                 visited[nr][nc][1] = visited[r][c][broken]+1
                que.push([nr, nc, broken+1])
            }
            
        }
        
    }
    console.log(answer)

})
/*
4
HELLO
DRINK
SHUTTLE
ZOO
*/